### PR TITLE
MED-296: Eliminate N+1 queries in task list hierarchy

### PR DESCRIPTION
**Pull request by JieyaZ**
## Changes
- Prefetch parent relationships for the current task list page.
- Preserve the existing API response.
- Add query-budget tests, five edge cases, and a parent-relationship payload snapshot.

### CODE REVIEW

File: frontend/e2e/tasks/tasks-helpers.ts [hunk 1]
```diff
@@ -73,7 +73,7 @@ export async function waitForTasksPageReady(page: Page) {
 export async function getActiveProjectSlug(page: Page): Promise<string> {
   const slug = await page.evaluate(() => {
     try {
-      const raw = localStorage.getItem('project-storage');
+      const raw = localStorage.getItem('project-storage-v1');
```
[the remark this answers is not in the document]
Comment (JieyaZ) [reply frontend/e2e/tasks/tasks-helpers.ts]: Fixed in 751762afd. All three helpers now read project-storage-v1 first and fall back to project-storage if the new key is missing. ESLint passed; Playwright was not rerun for this update.

File: backend/task/views.py [hunk 2]
```diff
@@ -193,6 +203,22 @@ def get_queryset(self):
                 TaskPin.objects.filter(task_id=OuterRef('pk'), user=user)
             ),
         )
+        if getattr(self, 'action', None) == 'list':
```
[the remark this answers is not in the document]
Comment (JieyaZ) [reply backend/task/views.py]: Fixed in 751762afd. The prefetch now runs only when include_subtasks=true. Added tests for the default, false, true, and uppercase TRUE values to verify the query behaviour. All 12 targeted backend tests passed.

**Review by quanwangniuniu — CHANGES_REQUESTED** [review]
## **Problem 0: Unnecessary Query on Default Path (Query Budget Regression)**

**Reason Why:**
The prefetch is currently gated only by checking `include_subtasks=true`, but the code that moves the prefetch retrieval into views.py (line 206-223) evaluates `include_subtasks` fresh on every request. However, there's a **timing issue**: when `include_subtasks=false` (the default), the prefetch still runs because it's added to the queryset unconditionally in the first pass, even though `get_parent_relationship()` will immediately return `None` for non-subtasks and ignore the prefetch.

Looking at line 208 in views.py:
```python
if getattr(self, 'action', None) == 'list' and include_subtasks:
```

This condition is correct, but the author's own test on line 21 (`test_parent_prefetch_only_runs_when_subtasks_requested`) has already confirmed this was originally failing and was then fixed. The fix is present in the current code, so this is actually **resolved** (the author addressed the Copilot reviewer's first concern).

---

## **Problem 1 (Actual): Potential Race Condition in E2E Storage Migration**

**Reason Why:**
The E2E helper functions in `frontend/e2e/tasks/tasks-helpers.ts` now read `project-storage-v1` first with a fallback to `project-storage`:
```typescript
const raw = localStorage.getItem('project-storage-v1') ?? localStorage.getItem('project-storage');
```

However, during the ongoing `project-storage` → `project-storage-v1` migration, there's a **narrow window where**:
1. One test writes to `project-storage-v1` synchronously  
2. Another test or browser tab still has only `project-storage` in localStorage  
3. The fallback works, but creates **flakiness if the migration is incomplete** across page reloads or in parallel test runs

**Potential Fix:**
Ensure the fallback chain is robust during migration by:
- Adding explicit retry logic in E2E helpers if `project-storage-v1` returns incomplete data
- Or guarantee that ALL tests hydrate both keys before assertions
- Consider adding a migration verification step in the test setup

---

## **Problem 2: Query Budget Assertion Too Lenient**

**Reason Why:**
In `test_task_list_hierarchy_edge_cases()` (line 79), the query count assertion is:
```python
assert len(queries) <= 12, [q['sql'] for q in queries]
```

A limit of **12 queries** for a paginated task list with hierarchy lookups is relatively loose. This makes it:
- Hard to detect **regressions in future changes** to the hierarchy prefetch logic
- Possible to hide additional N+1 queries introduced by unrelated code changes
- Not tight enough to guarantee the performance improvement is maintained

The stricter test `test_task_list_prefetches_parent_hierarchy_within_query_budget()` uses the same `<= 12` limit across all child counts (1, 19, 199), which means:
- A list of 199 subtasks should ideally be **~7-8 queries**, not 12
- If someone adds an accidental join or select_related() later, the regression won't be caught

**Potential Fix:**
Tighten the query budget with parameterized assertions:
```python
# For include_subtasks=true with N children:
# Expected: 1 (auth) + 1 (base tasks) + 1 (prefetch parent_relationships) + overhead = ~4-5 queries max
expected_max = 5 if include_subtasks else 2
assert len(hierarchy_queries) <= expected_max, f"Query budget exceeded: {len(hierarchy_queries)} > {expected_max}"
```

This ensures the optimization stays effective and catches performance regressions early.

**Comment by JieyaZ** [comment]
Thanks for the review. Problems 0–2 have been addressed: the prefetch is limited to requests that include subtasks, storage reads now handle incomplete data with fallback and retries, and query-budget assertions are stricter.
I also fixed the previous CI failure by restarting the frontend after the verification build and waiting for it to be ready before running Budget E2E. The latest CI run has passed.

**Review by quanwangniuniu — APPROVED** [review]
Good job
